### ISSUE TITLE
BLAS : Bad memory unallocation! :   50  0xffff12000000

### ISSUE THREAD

**Issue by erling-d-andersen** [xianyi/OpenBLAS#3470]
Why does OpenBLAS print out 

BLAS : Bad memory unallocation! :   50  0xffff12000000

Is it a serious issue.

**Comment by martin-frbg** [comment]
This certainly suggests something went wrong with thread memory management, though it does not necessarily mean that  the calculation failed. Which version of OpenBLAS is it (current is 0.3.18, or the `develop` branch on github), and on what hardware ?

**Comment by giaf** [comment]
Hi Martin, thanks for the quick reply.

Here are some more details:
- OpenBLAS dev version from 8a87e80c742a146d3900d64046d4f4c7fd58b6b0
- compiled on linux on aarch64 with gcc 9.2.1 toolchain
- key Makefile.rule options: USE_THREAD=0, USE_LOCKING=1 (the library is compiled as for a single thread but used in a multi-threaded algorithm)
- the issue does not seem to be reproducible, it seemingly happens randomly, and the likelihood of it happening increases with the number of threads used in the algorithm

If you need any more information, I will be happy to provide.

**Comment by martin-frbg** [comment]
How many threads in the algorithm ? You could try to make the message go away by building OpenBLAS with NUM_BUFFERS set to at least this number, the default is based on the number of cores seen in the build host (although recent versions should be able to grow their memory management tables as needed)

**Comment by giaf** [comment]
OpenBLAS was built on a machine with 4 cores, and the issue happened when running on another machine when using 32 or 64 threads.
The difficulty is that, the aim is to distribute the binaries, so there is no control on how many threads the end user will employ.

**Comment by martin-frbg** [comment]
Then I guess the issue is not happening with 32 threads (NUM_BUFFERS has a fixed minimum of 50), and there may be a bug in the recent code that switches between the management structures allocated at compile time and the new auxiliary set that gets added when the compile-time limit is exceeded.

**Comment by giaf** [comment]
The issue did happen also when 32 threads were used in the algorithm, in which case the printout was
```
BLAS : Bad memory unallocation! :   50  0xffff12000000
```
I will double check to see if by any chance more than 32 threads were somehow created in the algorithm, but in theory it was supposed to have exactly 32 threads.

**Comment by erling-d-andersen** [comment]
FYI:

We call OpenBlas from Intel TBB.  We compile it ourselves from source. Intel does not provide it.

(MKL has a single threaded sequential version i.e. you can call it from multiple threads and it is thread safe.)

**Comment by martin-frbg** [comment]
OpenBLAS is not aware of TBB in the sense that it handles interoperation with OpenMP, not sure if that matters here.
However this reminds me of #2760 (same thing on ppc, original bug resolved but issue kept open due to other suggestions made there). Maybe your hardware being ARM is the decisive factor - like ppc this has weak memory ordering so maybe the blas_lock function needs an ISB in its ARM64 counterpart to prevent a (relatively unlikely) race.

**Comment by erling-d-andersen** [comment]
OpenBLAS should just be thread safe like the sequential version of MKL is.  We cannot allow our BLAS library to run parallel
because we must parallelize ourselves. In any case OpenMP used not to be run-to-run deterministic. That is why we gave up on that years ago. Things may have changed in OpenMP since then but I prefer TBB to OpenMP.

I cannot say what is wrong but I am pretty sure OpenBLAS has an issue when called from many threads on the aarch64.

Do you test calling dgemm from many threads simultaneously?

**Comment by giaf** [comment]
Just to complement what Erling said, this maybe or may not be a related issue, but there seem to be some race conditions going on also in other places in OpenBLAS, and as you say this may be due to the fact that the ARM memory model is weaker, so code developed and working fine on x86 shows issues in this case.

Namely, we found that when the multi-threaded algorithm calls level 3 routines from (sequential) OpenBLAS on aarch64, there are different results appearing at random places and iterations (and this already for a small number of threads such as 2 or 4; for 1 thread there are no issues).
This doesn't happen when using another BLAS implementation altogether, nor also when OpenBLAS provides level 1 and 2 routines while another BLAS implementation provides level 3 routines (namely, the issue seem to happen in all of dgemm, dtrsm and dsyrk).
The original issue in this thread (i.e. the bad memory unallocation message) still happens in the case OpenBLAS provides only level 1 and 2 routines.

I wanted to open a separate issue about it once collecting some more information, but for now I wanted to mention it here too.

**Comment by erling-d-andersen** [comment]
On another problem I saw the stuff below. As commercial vendor using OpenBLAS I get really scared. 

BLAS : Program is Terminated. Because you tried to allocate too many memory regions.

**Comment by martin-frbg** [comment]
@giaf small test cases would be helpful, if possible. I do not have access to big multicore aarch64 except what is provided by the CI services, but I would expect anything involving only 2-4 cores should have shown up long ago. So far I have no idea what you may be doing differently - while the majority of the current userbase may be on x86_64 I doubt you are the only ones using OpenBLAS on aarch64.

@erling-d-andersen that message is what would happen when one exceeded the compile-time limit on the number of threads at runtime. (A design flaw in the original GotoBLAS of ~20 years ago, if you will). Are you still getting this with the current version (which should allocate a new control structure in this situation) ?
BTW if you turn this into a rant thread you will be blocked.

**Comment by erling-d-andersen** [comment]
Thanks for warning me. The intention was that the comments would be useful to someone and arm users in particular. And to confirm what we saw was unexpected. You comments have helped my understanding of Openblas. Thanks.

I think @giaf build Openblas from latest source.   He may have comments on that.

Ps The issue we have been seeing are very indeterministic and cannot be reproduced consistently. So creating a small example is likely  easier to be said than done.

**Comment by erling-d-andersen** [comment]
By mistake we have used 0.3.15. We will try 0.3.18 and report back.

**Comment by martin-frbg** [comment]
Expect no fundamental difference regarding the "bad unallocation", but as mentioned above 0.3.18 should expand its control structures instead of terminating when the compile-time thread limit is exceeded.
The "bad unallocation" was never seen with the DGEMM thread test (from cpp_thread_test folder) on drone.io (i.e. aarch64 with 48 and 92 cores) but unfortunately there is a problem with drone CI at the moment. If my hunch is correct, you could try adding
 `dmb  ish/n/t` before the `mov %0, #0` line in the inline assembly for blas_lock in common_arm64.h

**Comment by erling-d-andersen** [comment]
A quick test seems to indicate 0.3.18 works. However, we get multiple

OpenBLAS warning: precompiled NUM_THREADS exceeded, adding auxiliary array for thread metadata.

in some cases.

Is it possible to get rid of those? (We will ultimately get support calls if our customer sees those.)

**Comment by martin-frbg** [comment]
Build OpenBLAS with a large enough NUM_THREADS so that the problem does not arise in the first place (e.g. `make NUM_THREADS=512` or change the corresponding entry in Makefile.rule).

**Comment by giaf** [comment]
Hi Martin,
I have some updates after repeating the experiments with 0.3.18 (release version).
This seems to solve the issues regarding:
1) Rare wrong values with 2-4 threads
2) Bad memory unallocation. Even if you didn't expect changes in this regard, I could not reproduce this any longer, at least when trying out many more times than when I was getting the issue with 0.3.15. I will let you know if in the future it turns out to rise again, and the we can try your proposed fix in the lock.

However, there still seem to be issues whit the allocation of the new control structure.
More precisely, even if the library is compiled on a machine with 4 cores, all work fine when running up to 50 threads.
From 51, there are different combinations of the following warnings and crashes:
```
OpenBLAS warning: precompiled NUM_THREADS exceeded, adding auxiliary array for thread metadata.
OpenBLAS warning: precompiled NUM_THREADS exceeded, adding auxiliary array for thread metadata.
OpenBLAS : Program is Terminated. Because you tried to allocate too many memory regions.
This library was built to support a maximum of 50 threads - either rebuild OpenBLAS
OpenBLAS : Program is Terminated. Because you tried to allocate too many memory regions.
This library was built to support a maximum of 50 threads - either rebuild OpenBLAS
with a larger NUM_THREADS value or set the environment variable OPENBLAS_NUM_THREADS to
a sufficiently small number. This error typically occurs when the software that relies on
OpenBLAS calls BLAS functions from many threads in parallel, or when your computer has more
cpu cores than what OpenBLAS was configured to handle.
Segmentation fault (core dumped)
```
```
...
malloc(): unsorted double linked list corrupted
Aborted (core dumped)
```

```
...
malloc(): mismatching next->prev_size (unsorted)
Aborted (core dumped)
```
Setting NUM_THREADS at a large enough value at compilation time prevents the issue as no new control structure is allocated.

**Comment by brada4** [comment]
rare-wrong-values - how wrong? few bits significand loss is expected vs reference BLAS. if value fluctuates more than that on same (generated) input, pls share.

**Comment by martin-frbg** [comment]
Oops. That looks like a race involving the new flag for "are we in overflow already". Did not see that in my tests but they were mainly on x86_64. But increasing NUM_THREADS is definitely the way to go if you have the "luxury" of compiling OpenBLAS to fit the task. Everything else is more a stopgap solution for people stuck with whatever their distro came with (and where earlier OpenBLAS versions would have crashed and burned with the "tried to allocate too many" message anyway).

**Comment by erling-d-andersen** [comment]
@brada4: If rare-wrong-values refers to 0..3.15 then somewhat nondeterministic the results was completely messed up. It was like it wrote or read data arbitrary form memory in rare cases.

@martin-frbg: Compiling with a maximum of 512 threads is probably good enough for the near future. I do not know any computer with more than 512 core as I write this.

**Comment by brada4** [comment]
sounds as critical as you say.  if you find a repeater try to reduce it and share. causes can be many, on modern systems you cannot read (far) outside allocations, but you can read unitialized allocation in a long running process.

**Comment by erling-d-andersen** [comment]
I think we got the issue resolved, so I close.
